src/cpp/session/modules/SessionDependencies.R

```diff
@@ -0,0 +1,155 @@
+# this program is licensed to you under the terms of version 3 of the
+# GNU Affero General Public License. This program is distributed WITHOUT
+# ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+# MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+# AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+#
+#
+
+# Topographically sorts a list of packages (nodes) by their dependencies (edges). Note that this is
+# not meant to be a general-purpose topo sort algorithm, and that it returns packages in the correct
+# installation order, which is the exact reverse of traditional topological order. For example,
+# given packages with dependencies a -> b -> c, we would want to install packages in the order "c",
+# "b", and then "a".
+.rs.addFunction("topoSortPackages", function(nodes, edges) {
+   # List of sorted packages
+   sorted <- c()
+
+   # All nodes are unvisited to begin with
+   visited <- c()
+
+   # Define recursive descent function for dependencies
+   visit <- function(node, stack) {
+      if (node %in% visited) {
+         return()
+      }
+      if (node %in% stack) {
+         # We visited this node while visiting itself; this is a dependency loop.
+         stop("Package dependency graph is not a directed acyclic graph.")
+      }
+
+      # Visit all the edges of this node
+      stack <- c(stack, node)
+      for (edge in edges) {
+         if (identical(edge$from, node)) {
+            visit(edge$to, stack)
+         }
+      }
+
+      visited <<- c(visited, node)
+      sorted <<- c(sorted, node)
+   }
+
+   # Keep visiting unvisited nodes until we have visited all of them
+   while (length(visited) < length(nodes)) {
+      for (node in nodes) {
+         if (!(node %in% visited)) {
+            visit(node, c())
+            break
+         }
+      }
+   }
+
+   # Return topologically sorted list
+   sorted
+})
+
+.rs.addFunction("expandPkgDependencies", function(dependencies) {
+   .rs.expandDependencies(available.packages(), installed.packages(), dependencies)
+})
+
+.rs.addFunction("expandDependencies", function(available, installed, dependencies) {
+   # A list of nodes (package names) to be installed
+   nodes <- c()
+
+   # A list of details for packages to be installed
+   packages <- dependencies
+
+   # A vector of lists, with "from" and "to" named elements giving the dependencies
+   edges <- list()
+
+   # Get the dependencies of each package
+   for (dep in dependencies) {
+      # Add the package itself to the list of nodes
+      nodes <- c(nodes, dep$name)
+   }
+
+   # Look for dependencies of each package
+   for (dep in dependencies) {
+      # Dependencies are discovered from these three fields
+      fields <- c("Depends", "Imports", "LinkingTo")
+      for (field in fields) {
+         # Read contents for field (ignore if no contents)
+         contents <- available[dep$name, field]
+         if (!is.character(contents))
```

Comment:
I think we should test for `is.na()` here as well (since this could be `NA_character_`)

Reply:
Done!